test/modules/test_nf4_linear.py

```diff
@@ -0,0 +1,74 @@
+import logging
+import unittest
+
+import torch
+from torch import nn
+from torch.testing._internal.common_utils import TestCase
+from torchao.modules import FrozenNF4Linear
+from torchao.dtypes.nf4tensor import NF4Tensor
+
+bnb_available = False
+
+try:
+    import bitsandbytes as bnb
+    bnb_available = True
+except ImportError:
+    pass
+
+logging.basicConfig(
+    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s", level=logging.INFO
+)
+
+class TestNF4Linear(TestCase):
+    """
+    Test torchao.modules.NF4Linear
+    """
+    def test_bias_unsupported(self):
+        with self.assertRaisesRegex(RuntimeError, "does not currently support biases"):
+            _ = FrozenNF4Linear(1, 1, bias=True)
+
+    def test_non_bf16_unsupported(self):
+        with self.assertRaisesRegex(RuntimeError, "only supported with bf16"):
+            _ = FrozenNF4Linear(1, 1)
+
+    def test_frozen_nf4_linear(self):
+        nf4_linear = FrozenNF4Linear(512, 512, device='cpu', dtype=torch.bfloat16)
+        self.assertTrue(isinstance(nf4_linear.weight, NF4Tensor))
+        self.assertEqual(torch.bfloat16, nf4_linear.weight.get_original_weight().dtype)
```

Comment:
If you keep both, won't that affect memory consumption? If the user wants both, they can decide to keep both around. Otherwise they could convert and re-assign an nn.Linear.weight Tensor like here
```
def apply_sparse(model):
    apply_fake_sparsity(model)
    for name, mod in model.named_modules():
        if isinstance(mod, torch.nn.Linear):
            mod.weight = torch.nn.Parameter(to_sparse_semi_structured(mod.weight))
```

Reply:
Oh, I see. Maybe a more standard API would be to support `to(torch.bfloat16)` or such?

Reply:
> If you keep both, won't that affect memory consumption

So IIUC we aren't keeping both here, but can verify via looking at the memory allocation after creating an instance of `FrozenNF4Linear`. 

`get_original_weight` actually runs the dequantization and restores the original weight - maybe a bit of a misnomer since the name sorta implies its stored somewhere and is just accessed.

Reply:
yeah maybe "build_original_weight"

Reply:
But can you actually restore the original weight? Hasn't some fidelity been lost after converting to nf4? Hence my suggestion to just overwrite `to_dtype` here.

Reply:
That makes sense, it does seem we lose fidelity and we can't get the exact original weight as intuitively expected - 
<img width="850" alt="image" src="https://github.com/pytorch-labs/ao/assets/8039770/a7fc962b-e20f-412a-944f-eb3d26a05f96">

(@drisspg - just checking my understanding is correct). 

So should we update NF4Tensor to get rid of `get_original_weight` and just have a `.to()` API? @drisspg, can this be done in a separate PR?

Reply:
That is Christian suggestions not mine lol, but regardless I think this PR markedly increases the testing coverage, and like you said if we want to do the full switch over to subclasses and do everything through torch dispatch I think it would make sense to do in a follow up PR, cc @cpuhrsch